src/optical_spi.c

```diff
@@ -2,143 +2,139 @@
-        // In testing, timeout always ended at the original value minus 7 or 8
-        uint32_t timeout = UINT32_MAX;
-        while ((timeout > 0) && (opt_spi_num_bytes < 3)) {
-            timeout--;
-        }
-        print("timeout = %u\n", timeout);
-
-        // If we didn't get 3 bytes
-        if (opt_spi_num_bytes < 3) {
-            //_delay_ms(5000); // long delay = timeout, used for testing
-            return 0;
-        }
-
-        // If we have received all 3 bytes now
-        uint32_t read_data = opt_spi_data;
-        // print ("Received optical data: 0x%06lx\n", read_data);
 
-        // Clear SPI request variables
-        opt_spi_num_bytes = 0;
-        opt_spi_data = 0;
-
-        return read_data;
-    }
-
-    // Silence missing return warning
-    return 0;
+// returns status of DATA_RDY pin
+uint8_t get_data_pin(void){
+    return OPT_DATA_PIN & _BV(OPT_DATA);
 }
 
-/*
-Sends a command to read data for the specified field number from PAY-Optical.
-*/
-void send_opt_spi_cmd(uint8_t field_num) {
-    // print("Sending optical command - field #%u\n", field_num);
+
+// sends 2 bytes to request reading
+// byte 1 = cmd_byte
+// byte 2 = well data
+void send_opt_spi_cmd(uint8_t cmd_opcode, uint8_t well_data) {
+    // print("Sending optical the command code %X\n", cmd_opcode);
 
     opt_spi_data = 0;
-    opt_spi_num_bytes = 0;
+    spi_frame_number = 0;
 
-    // Send the command to PAY-Optical to start reading data
+    // Send the command to PAY-Optical to start reading data    
     set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
-    uint8_t spi_tx = (0b11 << 6) | field_num;
-    send_spi(spi_tx);
+    send_spi(cmd_opcode);
     set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
 
-    // After this, PAY-Optical should set DATA_RDY high, which will trigger the
-    // interrupt below to receive the data
+    // wait for DATA_RDYn to go LOW, signaling optical is done reading byte 1
+    // --> aka loop until DATA_RDYn is no longer HIGH
+    while (get_data_pin() == 1);
+
+    // print("Sending optical the well data %X\n", well_data);
+    set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
+    send_spi(well_data);
+    set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
 }
 
 
-/*
-This interrupt is triggered when the DATA_RDY pin goes set_cs_high
-(set by PAY-Optical to indicate data is ready).
-The SPI transfer should already be in progress if we get this interrupt.
-*/
-ISR(INT1_vect) {
-    // print("INT1 - DATA_RDY\n");
+void receive_opt_data(uint8_t num_expected_bytes){
+    
+    // still expecting more data from SSM
+    while (num_expected_bytes > spi_frame_number){
 
-    // Get the next byte over SPI
-    set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
-    uint8_t new_byte = send_spi(0x00);
-    set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
+        // OPTICAL ready to send data
+        if (get_data_pin() == 1){
+            uint8_t new_opt_data = SPDR;
+            opt_spi_data = opt_spi_data<<8 | new_opt_data;
+
+            spi_frame_number++;
```

Comment:
Might want to delay for a small amount of time (e.g. 1ms) to give optical time to set its DATA_RDY high then back to low when it has data